graph/manifest.go

```diff
@@ -8,62 +8,106 @@ import (
 		if err != nil {
-			return nil, false, fmt.Errorf("invalid manifest digest from registry: %s", err)
+			// near impossible
+			logrus.Errorf("error calculating local digest during tag pull: %v", err)
+			return "", nil, false, err
 		}
-
-		dgstVerifier, err := digest.NewDigestVerifier(manifestDigest)
+	} else {
+		// verify the manifest against local ref
+		verifier, err := digest.NewDigestVerifier(dgst)
 		if err != nil {
-			return nil, false, fmt.Errorf("unable to verify manifest digest from registry: %s", err)
+			// There are not many ways this can go wrong: if it does, its
+			// fatal. Likley, the cause would be poor validation of the
+			// incoming reference.
+			return "", nil, false, fmt.Errorf("error creating verifier for local digest reference %q: %v", dgst, err)
 		}
 
-		dgstVerifier.Write(payload)
+		if _, err := verifier.Write(payload); err != nil {
+			return "", nil, false, fmt.Errorf("error writing payload to local digest reference verifier: %v", err)
+		}
 
-		if !dgstVerifier.Verified() {
-			computedDigest, _ := digest.FromBytes(payload)
-			return nil, false, fmt.Errorf("unable to verify manifest digest: registry has %q, computed %q", manifestDigest, computedDigest)
+		if !verifier.Verified() {
+			return "", nil, false, fmt.Errorf("verification against local digest reference %q failed", dgst)
 		}
+
+		localDigest = dgst
 	}
 
-	if utils.DigestReference(ref) && ref != manifestDigest.String() {
-		return nil, false, fmt.Errorf("mismatching image manifest digest: got %q, expected %q", manifestDigest, ref)
+	// verify against the remote digest, if available
+	if remoteDigest != "" {
+		if err := remoteDigest.Validate(); err != nil {
+			return "", nil, false, fmt.Errorf("error validating remote digest %q: %v", remoteDigest, err)
+		}
+
+		verifier, err := digest.NewDigestVerifier(remoteDigest)
+		if err != nil {
+			// There are not many ways this can go wrong: if it does, its
+			// fatal. Likley, the cause would be poor validation of the
+			// incoming reference.
+			return "", nil, false, fmt.Errorf("error creating verifier for remote digest %q: %v", remoteDigest, err)
+		}
+
+		if _, err := verifier.Write(payload); err != nil {
+			return "", nil, false, fmt.Errorf("error writing payload to remote digest verifier (verifier target %q): %v", remoteDigest, err)
+		}
+
+		if !verifier.Verified() {
+			return "", nil, false, fmt.Errorf("verification against remote digest %q failed", remoteDigest)
+		}
 	}
 
 	var manifest registry.ManifestData
 	if err := json.Unmarshal(payload, &manifest); err != nil {
-		return nil, false, fmt.Errorf("error unmarshalling manifest: %s", err)
+		return "", nil, false, fmt.Errorf("error unmarshalling manifest: %s", err)
 	}
 	if manifest.SchemaVersion != 1 {
-		return nil, false, fmt.Errorf("unsupported schema version: %d", manifest.SchemaVersion)
+		return "", nil, false, fmt.Errorf("unsupported schema version: %d", manifest.SchemaVersion)
+	}
+
+	// validate the contents of the manifest
```

Comment:
could we move this check to after the verification block? ideally we'd have unmarshal down there too but it seems we have to parse here in order to get the `manifest.Name`?